productivity-suite/piercing-library/src/message-bus.ts

```diff
@@ -0,0 +1,76 @@
+export type MessageHandler = {
+  eventName: string;
+  callback: (detail: any) => void;
+};
+
+export type DispatchOptions = { singleUse?: boolean };
+
+export interface MessageBus {
+  dispatch(eventName: string, detail?: any, options?: DispatchOptions): void;
+  listen(handler: MessageHandler): () => void;
+}
+
+export class BrowserMessageBus implements MessageBus {
+  lastMessagesMap = new Map<
+    string,
+    { detail: any; dispatchOptions?: DispatchOptions } | null
+  >();
+  handlers: MessageHandler[] = [];
+
+  dispatch(eventName: string, detail: any, dispatchOptions: DispatchOptions) {
+    const handlersForEvent = this.handlers.filter(
+      ({ eventName: handlerEventName }) => handlerEventName === eventName
+    );
+    handlersForEvent.forEach(({ callback }) => callback(detail));
+    const eventConsumed = handlersForEvent.length && dispatchOptions?.singleUse;
+    if (!eventConsumed) {
+      this.lastMessagesMap.set(eventName, {
+        detail,
+        ...(dispatchOptions ? { dispatchOptions } : {}),
+      });
+    }
+  }
+
+  listen(handler: MessageHandler) {
+    const lastMessage = this.lastMessagesMap.get(handler.eventName);
+    if (lastMessage) {
+      handler.callback(lastMessage.detail);
+      if (lastMessage.dispatchOptions?.singleUse) {
+        this.lastMessagesMap.set(handler.eventName, null);
+      }
+    }
+    this.handlers.push(handler);
+    return () => {
+      this.handlers = this.handlers.filter((h) => h !== handler);
+    };
+  }
+}
+
+// export class ServerSideMessageBus implements MessageBus {
+//   constructor(request: Request) {}
+//   dispatch(eventName: string, detail: any): void {
+//     throw new Error("Method not implemented.");
+//   }
+//   listen(eventName: string, callback: () => void) {
+//     return () => {};
+//   }
+// }
+
+type WithOptionalMessageBus<T> = T & { messageBus?: MessageBus };
+
+export function getBus(element: Element): MessageBus {
+  let tmpEl: WithOptionalMessageBus<Element> | null = element;
+  while (tmpEl && tmpEl.messageBus === undefined) {
+    tmpEl = tmpEl.parentElement;
+  }
+
+  if (tmpEl?.messageBus) return tmpEl.messageBus;
+
+  const gThis = globalThis as WithOptionalMessageBus<typeof globalThis>;
+  if (!gThis.messageBus) {
+    // TODO: on the server this needs to be the server side message bus!
+    gThis.messageBus = new BrowserMessageBus();
+  }
+
+  return gThis.messageBus;
+}
```

Comment:
How about something like this?

```ts
export function getBus(element: Element): MessageBus {
  let tmpEl: Element | null = element;
  while (tmpEl) {
    if (hasMessageBus(tmpEl)) {
      return tmpEl.messageBus;
    }
    tmpEl = tmpEl.parentElement;
  }
  return ensureMessageBus(globalThis).messageBus;
}

function hasMessageBus(obj: unknown): obj is { messageBus: MessageBus } {
  return !!obj && "messageBus" in obj;
}

function ensureMessageBus<T extends object>(
  obj: T
): T & { messageBus: MessageBus } {
  if (hasMessageBus(obj)) {
    return obj;
  } else {
    const objWithBus = obj as T & { messageBus: MessageBus };
    objWithBus.messageBus = new BrowserMessageBus();
    return objWithBus;
  }
}
```

Reply:
Actually let's go a step further and throw an error if globalThis does not contain a MessageBus.

Reply:
```ts
export function getBus(element: Element): MessageBus {
  let tmpEl: Element | null = element;
  while (tmpEl) {
    if (hasMessageBus(tmpEl)) {
      return tmpEl.messageBus;
    }
    tmpEl = tmpEl.parentElement;
  }
  if (hasMessageBus(globalThis)) {
    return globalThis.messageBus;
  }
  throw new Error("No global message bus defined!");
}

function hasMessageBus(obj: unknown): obj is { messageBus: MessageBus } {
  return !!obj && "messageBus" in obj;
}
```

Reply:
I am not sure I follow, if `globalThis` doesn't have the message bus shouldn't we just create it? (isn't that how it gets created in the first place?)

Reply:
We will need to initialize the message bus with events from SSR, so I think it is cleaner to create it in a script that is injected into the SSR response.

Reply:
I've applied your latest suggestion 🙂👍